instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go

```diff
@@ -0,0 +1,132 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otelaws
+
+import (
+	"context"
+
+	v2Middleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+
+	"go.opentelemetry.io/contrib"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/label"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	instrumentationName = "go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-sdk-go-v2/otelaws"
+)
+
+type config struct {
+	TracerProvider trace.TracerProvider
+	Propagators    propagation.TextMapPropagator
+}
+
+// Option specifies instrumentation configuration options.
+type Option func(*config)
+
+// WithPropagators specifies propagators to use for extracting
+// information from the HTTP requests. If none are specified, global
+// ones will be used.
+func WithPropagators(propagators propagation.TextMapPropagator) Option {
+	return func(cfg *config) {
+		cfg.Propagators = propagators
+	}
+}
+
+// WithTracerProvider specifies a tracer provider to use for creating a tracer.
+// If none is specified, the global provider is used.
+func WithTracerProvider(provider trace.TracerProvider) Option {
+	return func(cfg *config) {
+		cfg.TracerProvider = provider
+	}
+}
+
+// AppendOtelMiddlewares attaches otel middlewares to aws go sdk v2 for instrumentation.
+// Otel middlewares can be appended to either all aws clients or a specific operation.
+// Please see more details in https://aws.github.io/aws-sdk-go-v2/docs/middleware/
+func AppendOtelMiddlewares(apiOptions *[]func(*middleware.Stack) error, opts ...Option) {
+	cfg := config{}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	if cfg.TracerProvider == nil {
+		cfg.TracerProvider = otel.GetTracerProvider()
+	}
+	if cfg.Propagators == nil {
+		cfg.Propagators = otel.GetTextMapPropagator()
+	}
+
+	awsTracer := cfg.TracerProvider.Tracer(
+		instrumentationName,
+		trace.WithInstrumentationVersion(contrib.SemVersion()),
+	)
+
+	*apiOptions = append(*apiOptions,
+		func(stack *middleware.Stack) error {
```

Comment:
Are these declared inline simply to close over `awsTracer`?  It looks like that's the only thing used from the containing scope, so I would suggest moving these to methods on a type that can hold on to a reference to the `Tracer` as it's a little hard to tease apart all the levels here.

```go
type mw struct { // this is probably a terrible type name, but it conflicts with the middleware package and I couldn't think of a better one :)
	tracer trace.Tracer
}

func (m mw) initialize(stack *middleware.Stack) error {...}
func (m mw) deserialize(stack *middleware.Stack) error {...}
```

Then registering the options becomes much cleaner: 

```go
m := mw{tracer: cfg.TracerProvider.Tracer(...)}
awsConfig.APIOptions = append(awsConfig.APIOptions, m.initialize, m.deserialize)
```

Reply:
good point, I inline 2 middlewares just for the scope of tracer, will update code to type.